File: javascript/146-LRU-Cache.js
/** 
 * https://leetcode.com/problems/lru-cache/
 * Time O(1) | Space O(N)
 * Your LRUCache object will be instantiated and called as such:
 * var obj = new LRUCache(capacity)
 * var param_1 = obj.get(key)
 * obj.put(key,value)
 */
 class LRUCache {
    constructor(capacity) {
        this.capacity = capacity;
        this.map = new Map();

        this.head = {};
        this.tail = {};

        this.head.next = this.tail;
        this.tail.prev = this.head;
    }

    removeLastUsed () {
        const [ key, next, prev ]  = [ this.head.next.key, this.head.next.next, this.head ];

        this.map.delete(key);
        this.head.next = next;
        this.head.next.prev = prev;
    }

    put (key, value) {
        const hasKey = this.get(key) !== -1;
        const isAtCapacity = this.map.size === this.capacity;
        
        if (hasKey) return (this.tail.prev.value = value);
        if (isAtCapacity) this.removeLastUsed();

        const node = { key, value };
        this.map.set(key, node);
        this.moveToFront(node);
    }

    moveToFront (node) {
        const [ prev, next ] = [ this.tail.prev, this.tail ];

        this.tail.prev.next = node;
        this.connectNode(node, { prev, next });
        this.tail.prev = node;
    }

    connectNode (node, top) {
        node.prev = top.prev;
        node.next = top.next;
    }

    get (key) {
        const hasKey = this.map.has(key);
        if (!hasKey) return -1;

        const node = this.map.get(key);
        
        this.disconnectNode(node);
        this.moveToFront(node);

        return node.value;
    }

    disconnectNode (node) {
        node.next.prev = node.prev;
        node.prev.next = node.next;
    }
}

File: go/LinkedList/146_LRU_Cache/146_LRU_Cache.go
package lrucache

import "fmt"

type Node struct {
	Key  int
	Val  int
	Prev *Node
	Next *Node
}

type LRUCache struct {
	cache    map[int]*Node
	capacity int
	Left     *Node
	Right    *Node
}

func Constructor(capacity int) LRUCache {
	cache := make(map[int]*Node)
	leftNode := &Node{}
	rightNode := &Node{}
	leftNode.Next, rightNode.Prev = rightNode, leftNode
	lruCache := LRUCache{
		cache:    cache,
		capacity: capacity,
		Left:     leftNode,
		Right:    rightNode,
	}
	return lruCache
}

func (this *LRUCache) remove(node *Node) {
	if node == nil {
		return
	}
	prev, next := node.Prev, node.Next
	prev.Next, next.Prev = next, prev
}

func (this *LRUCache) insert(node *Node) {
	prev, next := this.Right.Prev, this.Right
	prev.Next = node
	next.Prev = node
	node.Next, node.Prev = next, prev
}

func (this *LRUCache) Get(key int) int {
	val, ok := this.cache[key]
	if ok {
		this.remove(val)
		this.insert(val)
		// PrintList(this.Left)
		return val.Val
	}
	return -1
}

func (this *LRUCache) Put(key int, value int) {
	_, ok := this.cache[key]
	if ok {
		this.remove(this.cache[key])
	}
	this.cache[key] = &Node{
		Key: key,
		Val: value,
	}
	this.insert(this.cache[key])
	if len(this.cache) > this.capacity {
		lru := this.Left.Next
		delete(this.cache, lru.Key)
		this.remove(lru)

	}
	// PrintList(this.Left)
}

func PrintList(node *Node) {
	fmt.Print("List L -> R ")
	for node != nil {
		fmt.Print(node.Key, " ")
		node = node.Next
	}
	fmt.Println(" ")
}
